xdsl/dialects/riscv.py

```diff
@@ -0,0 +1,765 @@
+        "fp": 8,
+        "s0": 8,
+        "s1": 9,
+        "a0": 10,
+        "a1": 11,
+        "a2": 12,
+        "a3": 13,
+        "a4": 14,
+        "a5": 15,
+        "a6": 16,
+        "a7": 17,
+        "s2": 18,
+        "s3": 19,
+        "s4": 20,
+        "s5": 21,
+        "s6": 22,
+        "s7": 23,
+        "s8": 24,
+        "s9": 25,
+        "s10": 26,
+        "s11": 27,
+        "t3": 28,
+        "t4": 29,
+        "t5": 30,
+        "t6": 31
+    }
+
+    @staticmethod
+    def from_index(index: int) -> Register:
+        assert 32 > index >= 0
+        register = Register(index)
+        return register
+
+    @staticmethod
+    def from_name(name: str) -> Register:
+        if name in Register.abi_names:
+            return Register.from_index(Register.abi_names[name])
+        if name[0] == 'x' and name[1:].isnumeric():
+            return Register.from_index(int(name[1:]))
+        assert False, "register with unknown name"
+
+    def get_abi_name(self) -> str:
+        for name, index in Register.abi_names.items():
+            if index == self.index:
+                return name
+        assert False, "Register with unknown index"
+
+
+@irdl_attr_definition
+class RegisterAttr(Data[Register]):
+    name = "riscv.reg"
+
+    @staticmethod
+    def parse_parameter(parser: Parser) -> Register:
+        name = parser.parse_while(lambda x: x != ">")
+        return Register.from_name(name)
+
+    @staticmethod
+    def print_parameter(data: Register, printer: Printer) -> None:
+        printer.print_string(data)
+
+    @staticmethod
+    @builder
+    def from_index(index: int) -> RegisterAttr:
+        return RegisterAttr(Register.from_index(index))
+
+    @staticmethod
+    @builder
+    def from_name(name: str) -> RegisterAttr:
+        return RegisterAttr(Register.from_name(name))
+
+    @staticmethod
+    @builder
+    def from_register(register: Register) -> RegisterAttr:
+        return RegisterAttr(register)
+
+
+@irdl_attr_definition
+class LabelAttr(Data[str]):
+    name = "riscv.label"
+
+    @staticmethod
+    def parse_parameter(parser: Parser) -> str:
+        data = parser.parse_while(lambda x: x != ">")
+        return data
+
+    @staticmethod
+    def print_parameter(data: str, printer: Printer) -> None:
+        printer.print_string(data)
+
+    @staticmethod
+    @builder
+    def from_str(name: str) -> LabelAttr:
+        return LabelAttr(name)
+
+
+Op = TypeVar("Op", bound=Operation)
+
+
+class Riscv1Rd1Rs1ImmOperation(Operation):
+    rd: OpAttr[RegisterAttr]
+    rs1: OpAttr[RegisterAttr]
+    immediate: OpAttr[IntegerAttr]
+
+    @classmethod
+    def get(cls: Type[Op], rd, rs1, immediate) -> Op:
+        if isinstance(immediate, int):
+            immediate = IntegerAttr.from_int_and_width(immediate, 64)
+        return cls.build(attributes={
+            "rd": rd,
+            "rs1": rs1,
+            "immediate": immediate
+        })
+
+
+class Riscv1Rd1Rs1OffOperation(Operation):
```

Comment:
I find it somewhat foot-gunn-y to have helper classes that should not be instantiated here. What makes it even worse is the fact that instantiating them using `get` or `build` just returns `None`. Furthermore, marking them `ABC` does nothing to prevent that. I think the xDSL interface is quite lacking in this regard.

This is not a problem with this PR, but more an xDSL thing that probably needs improvement?

Reply:
Yeah I agree, I've been thinking about this quite a bit recently. I have some ideas, would be good to sync on what's been considered and rejected when Mathieu gets back. Writing the tutorials definitely helps to see what are the trickiest parts to explain, and the way to construct and not construct `Operation`s is definitely one of the hardest aspects of xDSL to start with